tests/system/Security/SecurityCSRFSessionRandomizeTokenTest.php

```diff
@@ -0,0 +1,263 @@
+use CodeIgniter\Session\Handlers\ArrayHandler;
+use CodeIgniter\Session\Session;
+use CodeIgniter\Test\CIUnitTestCase;
+use CodeIgniter\Test\Mock\MockAppConfig;
+use CodeIgniter\Test\Mock\MockSecurity;
+use CodeIgniter\Test\Mock\MockSession;
+use CodeIgniter\Test\TestLogger;
+use Config\App as AppConfig;
+use Config\Logger as LoggerConfig;
+use Config\Security as SecurityConfig;
+
+/**
+ * @runTestsInSeparateProcesses
+ * @preserveGlobalState disabled
+ *
+ * @internal
+ */
+final class SecurityCSRFSessionRandomizeTokenTest extends CIUnitTestCase
+{
+    /**
+     * @var string CSRF protection hash
+     */
+    private $hash = '8b9218a55906f9dcc1dc263dce7f005a';
+
+    /**
+     * @var string CSRF randomized token
+     */
+    private $randomizedToken = '8bc70b67c91494e815c7d2219c1ae0ab005513c290126d34d41bf41c5265e0f1';
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $_SESSION = [];
+        Factories::reset();
+
+        $config                 = new SecurityConfig();
+        $config->csrfProtection = Security::CSRF_PROTECTION_SESSION;
+        $config->tokenRandomize = true;
+        Factories::injectMock('config', 'Security', $config);
+
+        $this->injectSession($this->hash);
+    }
+
+    private function createSession($options = []): Session
+    {
+        $defaults = [
+            'sessionDriver'            => 'CodeIgniter\Session\Handlers\FileHandler',
+            'sessionCookieName'        => 'ci_session',
+            'sessionExpiration'        => 7200,
+            'sessionSavePath'          => null,
+            'sessionMatchIP'           => false,
+            'sessionTimeToUpdate'      => 300,
+            'sessionRegenerateDestroy' => false,
+            'cookieDomain'             => '',
+            'cookiePrefix'             => '',
+            'cookiePath'               => '/',
+            'cookieSecure'             => false,
+            'cookieSameSite'           => 'Lax',
+        ];
+
+        $config    = array_merge($defaults, $options);
+        $appConfig = new AppConfig();
+
+        foreach ($config as $key => $c) {
+            $appConfig->{$key} = $c;
+        }
+
+        $session = new MockSession(new ArrayHandler($appConfig, '127.0.0.1'), $appConfig);
+        $session->setLogger(new TestLogger(new LoggerConfig()));
+
+        return $session;
+    }
+
+    private function injectSession(string $hash): void
+    {
+        $session = $this->createSession();
+        $session->set('csrf_test_name', $hash);
+        Services::injectMock('session', $session);
+    }
+
+    public function testHashIsReadFromSession()
+    {
+        $security = new MockSecurity(new MockAppConfig());
+
+        $this->assertSame(
+            $this->randomizedToken,
+            $security->getHash()
+        );
+    }
+
+    public function testCSRFVerifyPostThrowsExceptionOnNoMatch()
+    {
+        $this->expectException(SecurityException::class);
```

Comment:
Include also the expected exception message.

Reply:
Added.